sphinx/source/docs/user_guide/data.rst

```diff
@@ -25,69 +30,122 @@ to the ``circle`` plotting method (see :ref:`userguide_plotting` for more exampl
+    cosine = np.cos(x)
+
+    p = figure()
+    p.circle(x=x, y=random)
+    p.line(x=x, y=cosine)
+
+.. _userguide_data_cds:
+
+Providing data as a ColumnDataSource
+------------------------------------
+
+The |ColumnDataSource| (CDS) is the core of most Bokeh plots. It provides the
+data to the glyphs of your plot.
+
+When you pass sequences like Python lists or NumPy arrays to a Bokeh renderer,
+Bokeh automatically converts those data structures into ColumnDataSources.
+Directly working with ColumnDataSources gives you access to more advanced
+options, compared to using standard data sequences like lists and arrays.
+
+For example: Using a |ColumnDataSource| allows you to share data between
+multiple plots and widgets. If you use a single ColumnDataSource together with
+multiple renderers, those renderers also share information about data you
+select with a select tool from Bokeh's toolbar (see
+:ref:`userguide_data_linked_selection`).
+
+Think of a ColumnDataSource as a collection of sequences of data that each have
+their own, unique column name.
+
+To create a |ColumnDataSource| object, you need a Python dictionary to pass to
+the object's ``data`` parameter:
+
+* Bokeh uses the dictionary's keys as column names.
+* The dictionary's values are used as the data values for your ColumnDataSource.
+
+The data you pass as part of your dict can be any non-string ordered sequences
+of values, such as lists or arrays (including NumPy arrays).
+
+Once you have a ColumnDataSource set up, pass it to a plotting function using
+the ``source`` argument:
 
 .. code-block:: python
 
     from bokeh.plotting import figure
     from bokeh.models import ColumnDataSource
 
+    # create a Python dict as the basis of your ColumnDataSource
     data = {'x_values': [1, 2, 3, 4, 5],
             'y_values': [6, 7, 2, 3, 6]}
 
+    # create a ColumnDataSource by passing the dict
     source = ColumnDataSource(data=data)
 
+    # create a plot using the ColumnDataSource's two columns
     p = figure()
     p.circle(x='x_values', y='y_values', source=source)
 
+To use a ColumnDataSource with a plotting function, you need to pass at least
+these three arguments:
+
+* ``x``: the name of the ColumnDataSource's column that contains the data for
+  the x values of your plot
+* ``y``: the name of the ColumnDataSource's column that contains the data for
+  the y values of your plot
+* ``source``: the name of the ColumnDataSource that contains the columns you
+  just referenced for the ``x`` and ``y`` arguments.
+
 .. note::
-    There is an implicit assumption that all the columns in a given ``ColumnDataSource``
-    all have the same length at all times. For this reason, it is usually preferable to
-    update the ``.data`` property of a data source "all at once".
+    Bokeh assumes that all columns in a ColumnDataSource each have the same
+    length at all times. For this reason, make sure to always update all columns
+    of a ColumnDataSource at the same time.
```

Comment:
there are things I don't think are good about both the old note and the new version. I'm not quite sure how to convy, though. The salient points:

* Lost (most?) usage of bokeh does not update a CDS after creation at all, so this is not relevant at all
* If you are creating a CDS, the columns all need the same length
* If you are adding a new column to `.data` then it needs to have the same length as the existing ones
* If you are updating data in a way that changes the column lengths, you need to update `.data` all at once (and I think it is good to be specific here about the technical step -- setting a new value for `.data`)

Reply:
I think it is at least important to mention what to keep in mind when updating an existing ColumnDataSource. So, how about this as the text for this note:

Bokeh assumes that all columns in a ColumnDataSource each have the same length at all times. For this reason, there are some things to keep in mind:

* When creating a new ColumnDataSource: Make sure all sequences of values you pass into a ColumnDataSource are the same length.
* When adding a new column to an existing ColumnDataSource by updating the ``.data`` property: Make sure your additional column is the same length as the columns already in the ColumnDataSource.
* When changing any values in an existing ColumnDataSource by updating the ``.data`` property: Make sure to either keep all columns the same length or set a completely new value for ``.data``, overwriting all data in the ColumnDataSource altogether.

Reply:
Closer, I'd really like to be very prescriptive. Not sure if it is evident but this was at times in the past a frequent source of support burden. 

* If you want to add a new column, you can do `source.data["newcol"] = some_array` as long as `some_array` is the same length as existing columns

* if you need to update data in such a way that the new data has different lengths, you must update all the columns all at once, by assigning an entire new `.data` dict

   ```
   source.data = { .... }
   ```
   or conversely, if you need to update to data with new lengths, you **cannot** do it by assignigning to one column at a time

Reply:
This is a lot of information for a "note"-section. I'll restructure this entire paragraph a bit, an new commit should be there shortly!